ecosystem/indexer-grpc/indexer-grpc-manager/src/metadata_manager.rs

```diff
@@ -0,0 +1,383 @@
+            client,
+            recent_states: VecDeque::new(),
+        }
+    }
+}
+
+struct HistoricalDataService {
+    client: DataServiceClient<Channel>,
+    recent_states: VecDeque<HistoricalDataServiceInfo>,
+}
+
+impl HistoricalDataService {
+    fn new(address: String) -> Self {
+        let channel = Channel::from_shared(address)
+            .expect("Bad address.")
+            .connect_lazy();
+        let client = DataServiceClient::new(channel);
+        Self {
+            client,
+            recent_states: VecDeque::new(),
+        }
+    }
+}
+
+pub(crate) struct MetadataManager {
+    chain_id: u64,
+    self_advertised_address: String,
+    grpc_managers: DashMap<String, Peer>,
+    fullnodes: DashMap<String, Fullnode>,
+    live_data_services: DashMap<String, LiveDataService>,
+    historical_data_services: DashMap<String, HistoricalDataService>,
+    known_latest_version: AtomicU64,
+}
+
+impl MetadataManager {
+    pub(crate) fn new(
+        chain_id: u64,
+        self_advertised_address: String,
+        grpc_manager_addresses: Vec<String>,
+        fullnode_addresses: Vec<String>,
+    ) -> Self {
+        let grpc_managers = DashMap::new();
+        for address in grpc_manager_addresses {
+            grpc_managers.insert(address.clone(), Peer::new(address));
+        }
+        let fullnodes = DashMap::new();
+        for address in fullnode_addresses {
+            fullnodes.insert(address.clone(), Fullnode::new(address));
+        }
+        Self {
+            chain_id,
+            self_advertised_address,
+            grpc_managers,
+            fullnodes,
+            live_data_services: DashMap::new(),
+            historical_data_services: DashMap::new(),
+            known_latest_version: AtomicU64::new(0),
+        }
+    }
+
+    fn need_ping(latest_state_timestamp: Timestamp, threshold: Duration) -> bool {
+        let latest_state_timestamp_since_epoch = Duration::new(
+            latest_state_timestamp.seconds as u64,
+            latest_state_timestamp.nanos as u32,
+        );
+        let now_since_epoch = SystemTime::now().duration_since(UNIX_EPOCH).unwrap();
+        let staleness = now_since_epoch.saturating_sub(latest_state_timestamp_since_epoch);
+
+        staleness >= threshold
+    }
+
+    pub(crate) async fn start(&self) -> Result<()> {
+        loop {
+            tokio_scoped::scope(|s| {
+                for kv in &self.grpc_managers {
+                    let grpc_manager = kv.value();
+                    let client = grpc_manager.client.clone();
+                    s.spawn(async move {
+                        let _ = self.heartbeat(client).await;
+                    });
+                }
+
+                for kv in &self.fullnodes {
+                    let (address, fullnode) = kv.pair();
+                    let need_ping = fullnode.recent_states.back().map_or(true, |s| {
+                        Self::need_ping(s.timestamp.unwrap_or_default(), Duration::from_secs(1))
```

Comment:
thought: config these time/duration; we may need to experiment these number

Reply:
I'm trying to avoid putting too many things into config, which confuse people. Let's keep it here now and only move it to config if we find it really necessary.